src/dotty/tools/dotc/transform/CheckIllegalOverrides.scala

```diff
@@ -0,0 +1,41 @@
+package dotty.tools.dotc
+package transform
+
+import TreeTransforms._
+import core._
+import Contexts._, Definitions._, Decorators._
+
+/** Checks if the user is trying to illegally override synthetic classes,
+  * primitives or their boxed equivalent
+  */
+class CheckIllegalOverrides extends MiniPhaseTransform {
+  import ast.tpd._
+
+  override def phaseName = "checkPhantom"
+
+  override def transformTypeDef(tree: TypeDef)(implicit ctx: Context, info: TransformerInfo) = {
+    val defn = ctx.definitions
+    val tpe = tree.tpe
+    val fullName = tree.symbol.showFullName
+
+    val syntheticClasses = defn.syntheticCoreClasses.map(_.showFullName)
+    val primitives = defn.ScalaValueClasses().map(_.showFullName)
+    val boxedPrimitives = defn.ScalaBoxedClasses().map(_.showFullName)
+
+    val illegalOverride =
+      if (syntheticClasses.contains(fullName))
+        Some("synthetic")
+      else if (primitives.contains(fullName))
+        Some("primitive")
+      else if (boxedPrimitives.contains(fullName))
+        Some("boxed primitive")
+      else
+        None
+
```

Comment:
It would be also nice to check that no-one tries to introduce a companion to those classes.
Eg, there actually is a class `Nothing$` that isn't source defined but backend relies on it existing.